subprojects/model-core/src/main/java/org/gradle/internal/instantiation/generator/AsmBackedClassGenerator.java

```diff
@@ -905,6 +911,29 @@ public void addPropertySetterOverloads(PropertyMetadata property, MethodMetadata
             }});
         }
 
+        @Override
+        public void addConfigurableFileCollectionSetterOverloads(PropertyMetadata property, MethodMetadata getter) {
+            if (!mixInDsl) {
+                return;
+            }
+
+            // GENERATE public void set<Name>(FileCollection p) {
+            //    ((ConfigurableFileCollection)<getter>()).setFrom(new Object[] { p });
+            // }
+            addSetter(getSetterName(property.getName()), RETURN_VOID_FROM_FILE_COLLECTION, methodVisitor -> new MethodVisitorScope(methodVisitor) {{
+                _ALOAD(0);
+                _INVOKEVIRTUAL(generatedType, getter.getName(), getMethodDescriptor(getType(getter.getReturnType())));
+                _CHECKCAST(CONFIGURABLE_FILE_COLLECTION_TYPE);
+                _ICONST_1();
+                _ANEWARRAY(OBJECT_TYPE);
+                _DUP();
+                _ICONST_0();
+                _ALOAD(1);
+                _AASTORE();
+                _INVOKEINTERFACE(CONFIGURABLE_FILE_COLLECTION_TYPE, "setFrom", ClassBuilderImpl.RETURN_VOID_FROM_OBJECT_ARRAY);
+            }});
+        }
```

Comment:
:question: Do we need a separate setter like this? Why not co-opt the existing setter taking an `Object` and adding a `setFromAnyValue()` to `DefaultConfigurableFileCollection` instead? With that approach we'd have more code in Java-land, and less in ASM-land, making it easier to maintain.

Reply:
Not sure I understand it correctly. Here we call `ConfigurableFileCollection.setFrom(Object...)`, which is already "generic". 

So if I understand you would want to simplify asm code by adding `setFromAnyValue()`, so we could just do something similar as for Property?

Reply:
We already have a method that adds `set<Property>(Object)` that calls `AbstractProperty.setFromAnyValue()`. I wonder if we could follow a similar pattern with `ConfigurableFileCollection`, maybe even reuse the existing method that adds the generated code instead of adding another method-generator.

What we could do is to move `PropertyInternal.setFromAnyValue()` to something like `LazyGroovySupport`, and let `PropertyInternal` and `DefaultConfigurableFileCollection` implement it. `LazyGroovySupport` could be moved to `model-groovy`, too, to make it obvious that it belongs to the Groovy support only.

Reply:
Ah, we can't move it to `model-groovy` and let `PropertyInternal` extend it at the same time... Nevertheless, I think moving this functionality out into its own interface would be good.

Reply:
Ok, maybe as a first step I can create `ConfigurableFileCollectionInternal` with `.setFromAnyValue` method and latter we consolidate a property and this new method. WDYT about that?

Reply:
Introducing `CFCInternal` sounds good. I'd still add `LazyGroovySupport`, though, and let `CFCInternal` extend it.